apis/r/tests/testthat/test-shape.R

```diff
@@ -1,107 +1,222 @@
+        expect_false(sdf$tiledbsoma_has_upgraded_domain())
+      } else {
+        expect_true(sdf$tiledbsoma_has_upgraded_domain())
+      }
+
+      # Check domain and maxdomain
+      dom = sdf$domain()
+      mxd = sdf$maxdomain()
+
+      # First check names
+      expect_equal(names(dom), index_column_names)
+      expect_equal(names(mxd), index_column_names)
+
+      # Then check all slots are pairs
+      for (name in names(dom)) {
+        expect_length(dom[[name]], 2L)
+        expect_length(mxd[[name]], 2L)
+      }
+
+      # Then check contents
+
+      # Old shape/domainishes (without core current domain) for non-string dims:
+      # * There is no core current domain
+      # * Expect domain == maxdomain
+      # * If they asked for NULL: both should be huge (near min/max for datatype)
+      # * If they asked for something specific: they should get it
+      #
+      # New shape/domainishes (with core current domain) for non-string dims:
+      # * Maxdomain should be huge (near min/max for datatype)
+      # * If they asked for NULL: domain should be the same as maxdomain
+      # * If they asked for a specific domain: they should get it
+      #
+      # Old shape/domainishes (without core current domain) for string dims:
+      # * There is no core current domain
+      # * Expect domain == maxdomain
+      # * Core domain for strings is always ("", "")
+      #
+      # New shape/domainishes (with core current domain) for string dims:
+      # * Core domain (soma maxdomain) for strings is always ("", "")
+      # * Core current domain (soma domain) for strings:
+      #   o If they asked for NULL: expect ("", "")
+      #   o If they asked for something specific: they should get it
+
+      if ("soma_joinid" %in% index_column_names) {
+        sjid_dom <- dom[["soma_joinid"]]
+        sjid_mxd <- mxd[["soma_joinid"]]
+        sjid_dfc <- domain_for_create[["soma_joinid"]]
+
+        if (!.new_shape_feature_flag_is_enabled()) {
+          # Old behavior
+          expect_equal(sjid_dom, sjid_mxd)
+        }
+
+        if (!use_domain_at_create) {
+          expect_equal(sjid_dom[[1]], 0)
+          expect_equal(sjid_mxd[[1]], 0)
+          # This is a really big number in the ballpark of 2**63; its exact
+          # value is unimportant.
+          expect_true(sjid_dom[[2]] > bit64::as.integer64(10000000000))
+          expect_true(sjid_mxd[[2]] > bit64::as.integer64(10000000000))
+        } else {
+          # The soma_joinid dim is always of type int64.  Everything coming back
+          # from libtiledbsoma, through C nanoarrow, through the R arrow
+          # package, to Arrow RecordBatch, holds true to that. But the final
+          # as.list() converts the domain to regular integer for us.  Not ideal
+          # IMO, but, we need to check against what it actually does.
+          #
+          # Not: expect_equal(sjid_dom, bit64::as.integer64(sjid_dfc))
+          expect_equal(sjid_dom, sjid_dfc)
```

Comment:
Downcasting can be prevented with `options(arrow.int64_downcast = FALSE)`; in a function that calls `as.list()` on an Arrow RecordBatch, put the following at the top of the method

```R
op <- options(arrow.int64_downcast = FALSE) # Prevent downcasting of int64 to integer or numeric
on.exit(options(op), add = TRUE, after = FALSE) # Restore previous configuration on function exit
```

Reply:
@mojaveazure the fix seems as weird as the problem -- what do you think of leaving the code as-is (but I'll update the comment)?

Reply:
I was just making a comment in case we wanted `sjid_dom` to be returned as a 64-bit integer. The "fix" is very normal in the R-world (I'd go so far as to say that we're the weird ones for using configuration/context objects instead of `options()`), and is already used several times in tiledbsoma-R
 - https://github.com/single-cell-data/TileDB-SOMA/blob/3103e816656f19e40f45f9cbeff920f345e98175/apis/r/R/IntIndexer.R#L29-L33
 - https://github.com/single-cell-data/TileDB-SOMA/blob/3103e816656f19e40f45f9cbeff920f345e98175/apis/r/R/utils.R#L123-L135
 - https://github.com/single-cell-data/TileDB-SOMA/blob/3103e816656f19e40f45f9cbeff920f345e98175/apis/r/R/SOMADataFrame.R#L90-L96
 - https://github.com/single-cell-data/TileDB-SOMA/blob/3103e816656f19e40f45f9cbeff920f345e98175/apis/r/R/BlockwiseIter.R#L184-L198

Reply:
@mojaveazure here is my proposal:
https://github.com/single-cell-data/TileDB-SOMA/pull/3032/commits/f0dfc4dfe62da733f9998046078e25363dccbd85

Reply:
Works for me